clippy_utils/src/msrvs.rs

```diff
@@ -124,62 +133,111 @@ impl Msrv {
+    /// nodes for that attribute. As such, prefer to run this check after cheaper pattern matching
+    /// operations.
     pub fn meets(self, cx: &LateContext<'_>, required: RustcVersion) -> bool {
         self.current(cx).is_none_or(|msrv| msrv >= required)
     }
 
+    /// Sets the current MSRV to the `rust-version` field specified in `Cargo.toml` if it is not
+    /// already set.
+    ///
+    /// If the current MSRV is set (usually through `clippy.toml`) and `rust-version` is also set
+    /// in `Cargo.toml`, this function will emit a warning if they are not the same value.
+    ///
+    /// This method is only meant to be called when creating `Conf`, and should not be called in
+    /// lint passes.
     pub fn read_cargo(&mut self, sess: &Session) {
         let cargo_msrv = std::env::var("CARGO_PKG_RUST_VERSION")
             .ok()
             .and_then(|v| parse_version(Symbol::intern(&v)));
 
         match (self.0, cargo_msrv) {
+            // If the MSRV is specified in `Cargo.toml` but not `clippy.toml`, use `Cargo.toml`'s
+            // value.
             (None, Some(cargo_msrv)) => self.0 = Some(cargo_msrv),
-            (Some(clippy_msrv), Some(cargo_msrv)) => {
-                if clippy_msrv != cargo_msrv {
-                    sess.dcx().warn(format!(
-                        "the MSRV in `clippy.toml` and `Cargo.toml` differ; using `{clippy_msrv}` from `clippy.toml`"
-                    ));
-                }
+
+            // If the MSRV is different in both `Cargo.toml` and `clippy.toml`, emit a warning and
+            // retain `clippy.toml`'s value.
+            (Some(clippy_msrv), Some(cargo_msrv)) if clippy_msrv != cargo_msrv => {
+                sess.dcx().warn(format!(
+                    "the MSRV in `clippy.toml` and `Cargo.toml` differ; using `{clippy_msrv}` from `clippy.toml`"
+                ));
             },
+
+            // If the MSRV is specified in `clippy.toml` or not at all, retain the current value.
             _ => {},
         }
     }
 }
 
-/// Tracks the current MSRV from `clippy.toml`, `Cargo.toml` or set via `#[clippy::msrv]` in early
-/// lint passes, use [`Msrv`] for late passes
+/// Tracks the current MSRV from `clippy.toml`, `Cargo.toml`, and `#[clippy::msrv]` in early lint
+/// passes.
+///
+/// This type is designed to be stored as a field of the early lint pass and updated with the
+/// [`extract_msrv_attr!`](crate::extract_msrv_attr) macro or by manually calling
+/// [`Self::check_attributes()`] and [`Self::check_attributes_post()`].
+///
+/// You can check the current MSRV with [`Self::current()`], and can check if it is equal or above
+/// a certain version with [`Self::meets()`].
+///
+/// If you are writing a late lint pass, use [`Msrv`] instead.
+///
+/// For more information, [see the development guide on
+/// MSRVs](https://doc.rust-lang.org/nightly/clippy/development/adding_lints.html#specifying-the-lints-minimum-supported-rust-version-msrv).
```

Comment:
The formatting seems a bit broken (though I think this is still valid Markdown)? Maybe something like this would help:
```suggestion
/// For more information, [see the development guide on MSRVs][doc-msrv].
/// 
/// [doc-msrv]: https://doc.rust-lang.org/nightly/clippy/development/adding_lints.html#specifying-the-lints-minimum-supported-rust-version-msrv
```